src/java.base/share/classes/sun/security/util/Hybrid.java

```diff
@@ -0,0 +1,435 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package sun.security.util;
+
+import com.sun.crypto.provider.DH;
+import sun.security.x509.X509Key;
+
+import javax.crypto.DecapsulateException;
+import javax.crypto.KEM;
+import javax.crypto.KEMSpi;
+import javax.crypto.SecretKey;
+import java.io.ByteArrayOutputStream;
+import java.math.BigInteger;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.InvalidParameterException;
+import java.security.Key;
+import java.security.KeyFactory;
+import java.security.KeyFactorySpi;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.KeyPairGeneratorSpi;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.ProviderException;
+import java.security.PublicKey;
+import java.security.SecureRandom;
+import java.security.spec.*;
+import java.util.Arrays;
+import java.util.Locale;
+
+public class Hybrid {
+
+    public record SecretKeyImpl(SecretKey k1, SecretKey k2) implements SecretKey {
+        @Override
+        public String getAlgorithm() {
+            return "Hybrid";
+        }
+
+        @Override
+        public String getFormat() {
+            return null;
+        }
+
+        @Override
+        public byte[] getEncoded() {
+            return null;
+        }
+    }
+
+    private static AlgorithmParameterSpec getSpec(String name) {
+        if (APS.isGenericEC(name)) {
+            return new ECGenParameterSpec(name);
+        } else {
+            return new NamedParameterSpec(name);
+        }
+    }
+
+    private static KeyPairGenerator getKeyPairGenerator(String name) throws
+            NoSuchAlgorithmException {
+        if (APS.isGenericEC(name)) {
+            name = "EC";
+        }
+        return KeyPairGenerator.getInstance(name);
+    }
+
+    private static KeyFactory getKeyFactory(String name) throws
+            NoSuchAlgorithmException {
+        if (APS.isGenericEC(name)) {
+            name = "EC";
+        }
+        return KeyFactory.getInstance(name);
+    }
+
+    /**
+     * Returns a KEM instance for each side of the hybrid algorithm.
+     * For traditional key exchange algorithms, we use the DH-based KEM
+     * implementation provided by DH.PROVIDER.
+     * For ML-KEM post-quantum algorithms, we obtain a KEM instance
+     * using the given algorithm name.
+     */
+    private static KEM getKEM(String name) throws NoSuchAlgorithmException {
+        if (APS.isGenericEC(name) || APS.isXDH(name)) {
+            return KEM.getInstance("DH", DH.PROVIDER);
```

Comment:
May I get more information about the hard-coded provider DH.PROVIDER?  Could it be more general so that other KEM provider can also be used?

Reply:
the DH provider is there give a KEM-like face on the NIST curves, x25519/448, etc., mainly for the purposes of assisting the work done within the hybrid KEM.  The underlying work being done by the internal DH provider still ends up doing KeyAgreement, KeyPairGenerator and KeyFactory operations, and those should go through the usual provider selection process.